pkg/engine/loadtargets.go

```diff
@@ -82,99 +68,79 @@ func resolveSpec(i int, target kyvernov1.ResourceSpec, ctx engineapi.PolicyConte
 	var targetObjects []resourceInfo
 	namespace := target.Namespace
 	name := target.Name
 	policy := ctx.Policy()
-
 	// if it's namespaced policy, targets has to be loaded only from the policy's namespace
 	if policy.IsNamespaced() {
 		namespace = policy.GetNamespace()
 	}
-	apiResource, parentAPIResource, _, err := client.Discovery().FindResource(target.APIVersion, target.Kind)
+	gvk := target.APIVersion + "/" + target.Kind
+	group, version, kind, subresource := kubeutils.ParseKindSelector(gvk)
+	gvrss, err := client.Discovery().FindResources(group, version, kind, subresource)
 	if err != nil {
 		return nil, err
 	}
-
-	if namespace != "" && name != "" &&
-		!wildcard.ContainsWildcard(namespace) && !wildcard.ContainsWildcard(name) {
-		// If the target resource is a subresource
-		var obj *unstructured.Unstructured
-		var parentResourceGVR metav1.GroupVersionResource
-		subresourceName := ""
-		if kubeutils.IsSubresource(apiResource.Name) {
-			apiVersion := metav1.GroupVersion{
-				Group:   parentAPIResource.Group,
-				Version: parentAPIResource.Version,
-			}.String()
-			subresourceName = strings.Split(apiResource.Name, "/")[1]
-			obj, err = client.GetResource(context.TODO(), apiVersion, parentAPIResource.Kind, namespace, name, subresourceName)
-			parentResourceGVR = metav1.GroupVersionResource{
-				Group:    parentAPIResource.Group,
-				Version:  parentAPIResource.Version,
-				Resource: parentAPIResource.Name,
-			}
-		} else {
-			obj, err = client.GetResource(context.TODO(), target.APIVersion, target.Kind, namespace, name)
+	for _, gvrs := range gvrss {
+		dyn := client.GetDynamicInterface().Resource(gvrs.GroupVersionResource)
+		var sub []string
+		if gvrs.SubResource != "" {
+			sub = []string{gvrs.SubResource}
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to get target %s/%s %s/%s : %v", target.APIVersion, target.Kind, namespace, name, err)
-		}
-
-		return []resourceInfo{{unstructured: *obj, subresource: subresourceName, parentResourceGVR: parentResourceGVR}}, nil
-	}
-
-	if kubeutils.IsSubresource(apiResource.Name) {
-		apiVersion := metav1.GroupVersion{
-			Group:   parentAPIResource.Group,
-			Version: parentAPIResource.Version,
-		}.String()
-		objList, err := client.ListResource(context.TODO(), apiVersion, parentAPIResource.Kind, "", nil)
-		if err != nil {
-			return nil, err
-		}
-		var parentObjects []unstructured.Unstructured
-		for i := range objList.Items {
-			obj := objList.Items[i].DeepCopy()
-			if match(namespace, name, obj.GetNamespace(), obj.GetName()) {
-				parentObjects = append(parentObjects, *obj)
+		// we can use `GET` directly
+		if namespace != "" && name != "" && !wildcard.ContainsWildcard(namespace) && !wildcard.ContainsWildcard(name) {
+			var obj *unstructured.Unstructured
+			var err error
+			if namespace == "" {
+				obj, err = dyn.Get(context.TODO(), name, metav1.GetOptions{}, sub...)
```

Comment:
`namespace` cannot be empty after the first `if`.

Reply:
Good catch, fixed.